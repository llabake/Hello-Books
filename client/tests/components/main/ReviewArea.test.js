import React from 'react';

import {
  ReviewArea,
  mapDispatchToProps,
  mapStateToProps
}
  from '../../../components/main/ReviewArea';

import adminData from '../../mocks/userData';
import userData from '../../mocks/userData';


function setup() {
  const props = {
    dispatch: jest.fn(),
    user: adminData.admin,
    handleAddReview: jest.fn(),
    handleChange: jest.fn(),
    errors: {},
    reviewBook: jest.fn().mockImplementation(() => Promise.resolve())
  };

  const state = {
    errors: {},
    isValid: false,
    saving: false,
    content: '',
    caption: ''
  };
  const wrapper = shallow(<ReviewArea {...props} {...state} />);

  return {
    wrapper,
    props,
    state,
  };
}

describe('Review area component', () => {
  test('should render and not crash', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot();

  })


  test('should call handleAddReview when the form is submitted', () => {
    const { wrapper, props, state } = setup()
    const event = {
      preventDefault: jest.fn(),
      target: {}
    }

    wrapper.instance().handleAddReview(event)
    expect(props.reviewBook).toHaveBeenCalled()
  });


  test(
    'should show validation errors when required form fields are empty',
    () => {
      const { wrapper, props, state } = setup();
      const event = {
        preventDefault: jest.fn(),
        target: {
          name: 'content',
          value: ''
        }
      };
      wrapper.setState({
        content: '',
        caption: 'v'
      });
      wrapper.instance().handleChange(event)
      expect(wrapper.instance().state.saving)
      .toEqual(false);
      expect(wrapper.instance().state.errors.content)
        .toEqual(   ["content is required"]);
    }
  );

  test('mapDispatchToProps', () => {
    const dispatch = jest.fn();

    expect(mapDispatchToProps(dispatch)).toHaveProperty('reviewBook');

    const { reviewBook } = mapDispatchToProps(dispatch);
    reviewBook();
    expect(dispatch).toHaveBeenCalled();
  });


  test('mapStateToProps', () => {
    expect(mapStateToProps({ bookReducer: {} })).toHaveProperty('errors');
  });
})