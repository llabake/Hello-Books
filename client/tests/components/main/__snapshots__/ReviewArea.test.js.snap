// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Review area component should render and not crash 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <TextInput
      icon="rate_review"
      id="caption"
      name="caption"
      onChange={[Function]}
      placeholder="Caption"
      type="text"
      value=""
    />
    <TextAreaInput
      icon="rate_review"
      id="content"
      name="content"
      onChange={[Function]}
      placeholder="Content"
      value=""
    />
    <button
      className="waves-effect waves-light btn right"
      disabled={true}
      style={
        Object {
          "marginLeft": "36em",
          "width": "25%",
        }
      }
      type="submit"
    >
      Post Review
    </button>
  </form>
</div>
`;
